### PR TITLE
PORTIA-598 Normalize start urls

### PR DESCRIPTION
**Pull request by sagelliv**
The `StartUrlCollection` allows three types of start urls:

* `StringUrl`: legacy start url which is just a string
* `GeneratedUrl`: legacy start url which was used for kimono urls
* `StartUrl`: new start urls based on fragments

I exposed `normalize` in `StartUrlCollection` which normalizes all types of urls to use the new fragments structure. This allows us to normalize start urls in the `pre_load` hook in the ORM to start gradually converting start urls.

This normalization in turn simplifies the validations we perform on the start urls field for spiders using the ORM validators.

### CODE REVIEW

File: slybot/slybot/starturls/generated_url.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        base_url = self.spec['template'].format(*paths)
+
+        query_params = [normalize_url_query_param(path, is_first=(i == 0))
+                        for i, path in enumerate(self._query_params)]
+        return base_url + ''.join(query_params)
+
+    @property
+    def normalized_fragments(self):
+        fixed, path = self._fixed_fragments, self._path_fragments
+        fragments = zip(fixed, path)
```
Comment (ruairif) [inline slybot/slybot/starturls/generated_url.py]: Can you use `list(zip(fixed, path))` for python 3 compatability

File: slybot/slybot/starturls/generated_url.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        return list(chain(*fragments))
+
+    @property
+    def _query_params(self):
+        params = self.spec['params']
+        template_params = self.spec['params_template']
+
+        query_params = []
+        for query, value in template_params:
+            if query not in [p['name'] for p in params]:
```
Comment (ruairif) [inline slybot/slybot/starturls/generated_url.py]: Could you use:
```python
param_names = {p['name'] for p in params}
...
    if query not in param_names:
        ...
```

File: slybot/slybot/starturls/generated_url.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    return [first_fragment, normalize_path(x)]
+
+
+def normalize_default(x):
+    return str(x['values'][0])
+
+
+def normalize_range(x):
+    a, b = x['values'][0:2]
+    b_inclusive = str(int(b) - 1)
```
Comment (ruairif) [inline slybot/slybot/starturls/generated_url.py]: Should this be `str(int(b) + 1)`?

Comment (sagelliv) [inline slybot/slybot/starturls/generated_url.py]: In the `_process_range` function in the `FragmentGenerator` we are adding 1 so we need to subtract here.

**Pull request merged**